frontend/device/generic/device.lua

```diff
@@ -1024,26 +1025,36 @@ end
 function Device:unpackArchive(archive, extract_to, with_stripped_root)
     require("dbg").dassert(type(archive) == "string")
     local BD = require("ui/bidi")
-    local ok
-    if archive:match("%.tar%.bz2$") or archive:match("%.tar%.gz$") or archive:match("%.tar%.lz$") or archive:match("%.tgz$") then
-        ok = self:untar(archive, extract_to, with_stripped_root)
-    else
-        return false, T(_("Couldn't extract archive:\n\n%1\n\nUnrecognized filename extension."), BD.filepath(archive))
+    local arc = Archiver.Reader:new()
+    local ok = arc:open(archive)
+    if ok then
+        for entry in arc:iterate() do
+            local dest_path = entry.path
+            if with_stripped_root then
+                local __, tail = dest_path:match("([^/]*)/*(.*)")
+                if tail then
+                    -- Non-root: strip one level.
+                    dest_path = tail
+                elseif entry.mode == 'directory' then
+                    -- Root directory: ignore.
+                    goto continue
+                else -- luacheck: ignore 542
+                    -- Root non-directory: don't strip.
+                end
+            end
+            if not arc:extractToPath(entry.path, extract_to.."/"..dest_path) then
+                break
+            end
+            ::continue::
+        end
+        ok = not arc.err
     end
     if not ok then
-        return false, T(_("Extracting archive failed:\n\n%1"), BD.filepath(archive))
+        return false, T(_("Couldn't extract archive:\n\n%1\n\n%s."), BD.filepath(archive), arc.err)
```

Comment:
Typo (should be `%2` with this templating function), but we can probably just stick the other stuff in a concat outside of the translation.